libraries/Microsoft.Bot.Builder.LanguageGeneration/LGFile.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Bot.Expressions;
+using Microsoft.Bot.Expressions.Memory;
+
+namespace Microsoft.Bot.Builder.LanguageGeneration
+{
+    /// <summary>
+    /// LG entrance, including properties that LG file has, and evaluate functions.
+    /// </summary>
+    public class LGFile
+    {
+        private readonly ExpressionEngine expressionEngine;
+
+        public LGFile(
+            IList<LGTemplate> templates = null,
+            IList<LGImport> imports = null,
+            IList<Diagnostic> diagnostics = null,
+            IList<LGFile> references = null,
+            string content = null,
+            string id = null,
+            ExpressionEngine expressionEngine = null,
+            ImportResolverDelegate importResolver = null)
+        {
+            Templates = templates ?? new List<LGTemplate>();
+            Imports = imports ?? new List<LGImport>();
+            Diagnostics = diagnostics ?? new List<Diagnostic>();
+            References = references ?? new List<LGFile>();
+            Content = content ?? string.Empty;
+            Id = id ?? string.Empty;
+            this.expressionEngine = expressionEngine ?? new ExpressionEngine();
+            this.ImportResolver = importResolver;
+        }
+
+        public IList<LGTemplate> AllTemplates
+        {
+            get
+            {
+                var referenceTemplates = References.GroupBy(x => x.Id).Select(x => x.First()).SelectMany(x => x.Templates);
```

Comment:
why you need to groupby?

Reply:
another suggestion is to union first, union (this+references) then selectMany, looks a little more well organized.